### PR TITLE
SR-16/SR-124 Playwright framework setup and create inquiries

### PR DESCRIPTION
**Pull request by rashmiiyengar**
# Task

Refactored test structure by implementing the Page Object Model and relocating all .spec.ts files to the tests folder for better organization. Added inquiry.spec.ts to include all updates related to SR-16.

[Link to Ticket]
https://incredihire-academy.atlassian.net/jira/software/c/projects/SR/boards/3?selectedIssue=SR-16

[Insert related tickets here. Otherwise, delete this block]
https://incredihire-academy.atlassian.net/jira/software/c/projects/SR/boards/3?selectedIssue=SR-124

# Author Tasks

## Acceptance Criteria
As an Admin, I want to publish a set of Inquiries that will be presented to users, to collect details on users’ state of mind, etc., so that users and managers can align state of mind details with team and individual activities.  

ACCEPTANCE CRITERIA

Admin portal should have a button that allows the admin to add an inquiry

“Add Inquiry” Button should trigger a pop-up window that displays an input box labeled "Enter the text of your inquiry." 

Style the input box and question list to ensure a responsive and user-friendly interface.

error messages under the inputs for validation

toast notification for network or backend errors to cover the necessary feedback.

Pop-up window should have a SAVE and CANCEL button

The SAVE button closes the window and displays a success message

The CANCEL button or the X closes the window, and no inquiry is added to the database.

The system should store the inquiries in the database to prevent the entry of duplicate questions.

### CODE REVIEW

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    await this.page.waitForSelector(".chakra-modal__footer", {
+      state: "visible",
+    })
+    await expect(
+      this.page.getByLabel("Add Inquiry").getByText("Add Inquiry"),
+    ).toHaveText("Add Inquiry")
+  }
+
+  async submitInquiry() {
+    const randomQuestion = this.getRandomQuestion()
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:28]: As a general principle, it's not a good idea to have randomness in tests. The reason is that if some random value causes the test to fail, simply retrying the failed test can cause it to succeed, leading the author to think the test was simply flaky. I would recommend picking specific values, or maybe having multiple tests with multiple edge-case values.

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    await this.navigateToInquiriesPage()
+    await this.openInquiryForm()
+    await this.fillTextByTestId("add-inquiry-text", randomQuestion)
+    await this.assertButtonTextByTestIdAndClick("submit-add-inquiry", "Save")
+    await this.waitForNumberOfSeconds(3)
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:33]: this tends to be another smell in automated tests. When possible, try to avoid sleeps. (since when executed on different machines with potentially different CPU loads, the sleep can legitimately time out.) If Playwright has some affordance to "waitUntil" or something like that, I'd use that instead.

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    await this.page.getByRole("button", { name: "Inquiry" }).click()
+    await this.page.waitForSelector(".chakra-modal__footer", {
+      state: "visible",
+    })
+    await expect(
+      this.page.getByLabel("Add Inquiry").getByText("Add Inquiry"),
+    ).toHaveText("Add Inquiry")
+  }
+
+  async submitInquiry() {
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:27]: a good best practice to follow when naming your tests is:

<functionality>_when<Scenario>_should<ExpectedOutcome>

For example,

addInquiry_whenValidInquiryIsAdded_shouldShowInquiryInInquiriesList()

And so on.

Comment (Eugene-Polonsky) [reply frontend/e2e/page-objects/InquiryPage.ts:27]: (this is from Roy Osherove's Art of Unit Testing)

File: frontend/e2e/page-objects/helperBase.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import questionsData from "../test-data/questions.json" assert { type: "json" }
+
+class HelperBase {
+  readonly page: Page
+
+  constructor(page: Page) {
+    this.page = page
+  }
+
+  async waitForNumberOfSeconds(timeinSeconds: number) {
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/helperBase.ts:11]: yeah again, I'd try to avoid this.

File: frontend/e2e/page-objects/helperBase.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    await button.click()
+  }
+
+  async fillTextByTestId(testId: string, text: string) {
+    const textArea = this.page.locator(`[data-testid="${testId}"]`)
+    await textArea.waitFor({ state: "visible" })
+    await textArea.fill(text)
+  }
+
+  getRandomQuestion(): string {
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/helperBase.ts:36]: like I said, I wouldn't do this.

File: frontend/e2e/tests/inquiry.spec.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+test.describe("Inquiry Management Suite", () => {
+  test.beforeEach(async ({ page }) => {
+    await page.goto("/")
+  })
+
+  test.afterEach(async ({ page }) => {
+    await page.close()
+  })
+
+  test("TC_001 Verify the presence of AddInquiry button on Inquiry Management page", async ({
```
Comment (Eugene-Polonsky) [inline frontend/e2e/tests/inquiry.spec.ts:13]: for Jest-style tests, this different naming convention is totally fine. The key thing to think about is that a test should only ever fail for one reason, and that reason needs to be obvious from test name.

File: frontend/e2e/page-objects/InquiryPage.ts (hunk 1 above)
Comment (rashmiiyengar) [reply frontend/e2e/page-objects/InquiryPage.ts:28]: Thank you for your feedback! I understand the concern about having randomness in tests and how that can lead to flaky results. To address this, I’ve made changes to ensure that each test uses a fixed input value. 
I’ve removed the random selection of questions and replaced it with predefined inquiry text for each test. Each test case now independently verifies that an inquiry is added and displayed in the list. I hope this resolves the concern and ensures better stability for the tests.

File: frontend/e2e/page-objects/InquiryPage.ts (hunk 2 above)
Comment (rashmiiyengar) [reply frontend/e2e/page-objects/InquiryPage.ts:33]: Thank you for the feedback, I’ve removed the fixed sleep (waitForNumberOfSeconds) and replaced it with a more reliable method to wait for the inquiry to appear, instead using fixed time i now wait for selector to be visible

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    await this.page.waitForSelector(".chakra-modal__footer", {
+      state: "visible",
+    })
+    await expect(
+      this.page.getByLabel("Add Inquiry").getByText("Add Inquiry"),
+    ).toHaveText("Add Inquiry")
+  }
+
+  async addInquiry_whenValidInquiryIsAdded_shouldShowInquiryInInquiriesList(
+    inputText: string,
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:28]: question: so how does this test get executed by playwright? Would it automatically create different input text?

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      state: "visible",
+    })
+  }
+
+  async addInquiry(inputText: string) {
+    await this.fillTextByTestId("add-inquiry-text", inputText)
+    await this.assertButtonTextByTestIdAndClick("submit-add-inquiry", "Save")
+  }
+
+  async addInquiry_whenValidInquiryIsAdded_shouldShowInquiryInInquiriesList(
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:21]: Meh, sorry for the confusion. So this function should no longer be necessary, right? I.e. you can just call addInquiry directly from the spec?

File: frontend/e2e/page-objects/InquiryPage.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    inputText: string,
+  ) {
+    await this.addInquiry(inputText)
+  }
+
+  async inquiry_whenInvalidInquiryText_shouldShowValidationError(
+    inputText: string,
+  ) {
+    await this.addInquiry(inputText)
```
Comment (Eugene-Polonsky) [inline frontend/e2e/page-objects/InquiryPage.ts:30]: ditto here.

File: frontend/e2e/tests/inquiry.spec.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  }) => {
+    const pm = new PageManager(page)
+    await pm.oninquiryPage()
+    await pm.oninquiryPage().navigateToInquiriesPage()
+    await pm.oninquiryPage().openInquiryForm()
+
+    const inputText = "How would you rate the work environment in your team?"
+    await pm
+      .oninquiryPage()
+      .addInquiry_whenValidInquiryIsAdded_shouldShowInquiryInInquiriesList(
```
Comment (Eugene-Polonsky) [inline frontend/e2e/tests/inquiry.spec.ts:24]: so here i think you can just call addInquiry directly, right?

File: frontend/e2e/tests/inquiry.spec.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  }) => {
+    const pm = new PageManager(page)
+    await pm.oninquiryPage()
+    await pm.oninquiryPage().navigateToInquiriesPage()
+    await pm.oninquiryPage().openInquiryForm()
+
+    const inputText = ""
+    await pm
+      .oninquiryPage()
+      .inquiry_whenInvalidInquiryText_shouldShowValidationError(inputText)
```
Comment (Eugene-Polonsky) [inline frontend/e2e/tests/inquiry.spec.ts:42]: ditto here.

**Review by Eugene-Polonsky — APPROVED** [review]
